hibernate-core/src/main/java/org/hibernate/collection/spi/AbstractPersistentCollection.java

```diff
@@ -282,7 +283,11 @@ else if ( !session.isConnected() ) {
 				}
 			}
 			else {
-				if ( !session.isTransactionInProgress() ) {
+				/*
+				 Whenever the collection lazy loading is triggered while consuming a ResultSet the logical connection will be physical connected
+				 and closing the connection will cause an error to occur when RowProcessingStateStandardImpl#next() will be called.
+				 */
+				if ( !session.isTransactionInProgress() && !isPhysicallyConnected ) {
```

Comment:
Ok, as far as I understand, it is only valid to call `afterTransaction` when the connection was acquired through this method. In that case, wouldn't it be better to check whether the stack in `PersistenceContext#getLoadContexts` is empty, and only then run this? i.e. it is safe to run `afterTransaction` when no load contexts are available, as that means this method acquired a connection in the first place.

Reply:
`persistenceContext#getLoadContexts()` cannot be used because it creates one if null, so I used `session.getPersistenceContext().isLoadFinished()` instead, what do you think?